gramps/plugins/graph/gvhourglass.py

```diff
@@ -153,59 +160,92 @@ def traverse_down(self, person, gen):
                 # allocate only one father per family
                 self.__family_father.append(family_handle)
                 father = self.__db.get_person_from_handle(father_handle)
-                self.add_person(father)
+                self.add_person(father, fathersosanumber)
+                father_id = father.get_gramps_id()
                 self.doc.add_link(father.get_gramps_id(), family_id,
                                   head=self.arrowtailstyle,
                                   tail=self.arrowheadstyle )
+                # update node with its father node id
+                self.__node_label[person_id] = [self.__node_label[person_id][0], father_id, self.__node_label[person_id][2]]
                 # no need to go up if he is a father in another family
                 if father_handle not in self.__used_people:
                     self.__used_people.append(father_handle)
-                    self.traverse_up(father, gen+1)
+                    self.traverse_up(father, gen+1, fathersosanumber)
+            elif family_handle in self.__family_father and self.ahnentafelnum:
+                self.rewrite_sosa_number(self.__db.get_person_from_handle(father_handle).get_gramps_id(), fathersosanumber)
 
             # create link from family to mother
             mother_handle = family.get_mother_handle()
             if mother_handle and family_handle not in self.__family_mother:
                 # allocate only one mother per family
                 self.__family_mother.append(family_handle)
                 mother = self.__db.get_person_from_handle(mother_handle)
-                self.add_person(mother)
+                self.add_person(mother, mothersosanumber)
+                mother_id = mother.get_gramps_id()
                 self.doc.add_link(mother.get_gramps_id(), family_id,
                                   head=self.arrowtailstyle,
                                   tail=self.arrowheadstyle)
+                # update node with its mother node id
+                self.__node_label[person_id] = [self.__node_label[person_id][0], self.__node_label[person_id][1], mother_id]
                 # no need to go up if she is a mother in another family
                 if mother_handle not in self.__used_people:
                     self.__used_people.append(mother_handle)
-                    self.traverse_up(mother, gen+1)
+                    self.traverse_up(mother, gen+1, mothersosanumber)
+            elif family_handle in self.__family_mother and self.ahnentafelnum:
+                self.rewrite_sosa_number(self.__db.get_person_from_handle(mother_handle).get_gramps_id(), mothersosanumber)
+
+            if self.ahnentafel and mother_handle and father_handle and father_id != '' and mother_id != '':
+                self.doc.add_link(father_id, mother_id,
+                                  style='invis')
+                self.doc.add_samerank(father_id, mother_id)
+
+    def rewrite_sosa_number(self, pid, sosanumber):
```

Comment:
Why it is necessary to rewrite existing labels?  When traversing up the tree, if you find a person already in the tree, can't you just use their existing Sosa number and continue?

Maybe I don't fully understand the algorithm here.

Reply:
Because a person can have multiple sosa number.
A number 4, can also be a number 6 is his chidren have children between them